### PR TITLE
Fix invalid isolated inference of resource functions which has start actions

### LINKED ISSUE #41691: [Bug]: Invalid inferring of `isolated` with start actions when the arguments are not isolated expressions
### Description

$title.

### Steps to Reproduce

```ballerina
import ballerina/http;

type UserRegistration record {
    string name;
    string email;
};

service on new http:Listener(8080) {
    resource function get h1(UserRegistration user) {
        _ = start sendEmailNotification(user);
    }
}

isolated function sendEmailNotification(UserRegistration user) returns error? {
}
```

Here, it seems like the resource method is inferred as `isolated` - no warning - despite the argument not being an isolated expression.

But, if we explicitly mark the resource method as `isolated`, it results in a compilation error for the argument.

```ballerina
import ballerina/http;

type UserRegistration record {
    string name;
    string email;
};

service on new http:Listener(8080) {
    isolated resource function get h1(UserRegistration user) {
        _ = start sendEmailNotification(user);
    }
}

isolated function sendEmailNotification(UserRegistration user) returns error? {
}
```

```cmd
$ bal run --offline sample.bal 
Compiling source
        sample.bal
ERROR [sample.bal:(10:41,10:45)] invalid start action accessing a non isolated expression in an argument of a async call of an 'isolated' function
```

### Affected Version(s)

2201.8.0

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### LINKED ISSUE #41766: [Bug]: Compiler crash when record is given as rest arg for a method
### Description

$subject
Attempting to pass required arguments as rest parameters using a record value results in a compiler crash.

### Steps to Reproduce

```ballerina
type Record record {|
    string[] arr1;
    string[] arr2;
    string[] arr3;
|};

service on new Listener() {
    resource function post qux(Record rec) {
        _ = start f(...rec);
    }
}

isolated function f(string[] arr1, string[] arr2, string[] arr3) {
}

public isolated class Listener {
    public function init() {
    }

    public function attach(service object {} s, string[]|string|() name = ()) returns error? {
    }

    public function 'start() returns error? {
    }

    public function gracefulStop() returns error? {
    }

    public function immediateStop() returns error? {
    }

    public function detach(service object {} s) returns error? {
    }
}
```

### Affected Version(s)

2201.8.2

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### LINKED ISSUE #41818: [Bug]: Compiler crash when passing arguments where an empty list constrcutor
### Description

$subject

Compiling the below code will give a compiler crash.

```

[2023-12-04 11:14:01,690] SEVERE {b7a.log.crash} - Index 0 out of bounds for length 0 
java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
        at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
        at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
        at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:266)
        at java.base/java.util.Objects.checkIndex(Objects.java:361)
        at java.base/java.util.ArrayList.get(ArrayList.java:427)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.IsolationAnalyzer.analyzeArgIsolatedness(IsolationAnalyzer.java:2331)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.IsolationAnalyzer.analyzeInvocation(IsolationAnalyzer.java:2082)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.IsolationAnalyzer.visit(IsolationAnalyzer.java:1422)
        at org.wso2.ballerinalang.compiler.tree.expressions.BLangInvocation.accept(BLangInvocation.java:128)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.IsolationAnalyzer.analyzeNode(IsolationAnalyzer.java:322)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.IsolationAnalyzer.visit(IsolationAnalyzer.java:580)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangAssignment.accept(BLangAssignment.java:81)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.IsolationAnalyzer.analyzeNode(IsolationAnalyzer.java:322)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.IsolationAnalyzer.visit(IsolationAnalyzer.java:449)
        at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:60)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.IsolationAnalyzer.analyzeNode(IsolationAnalyzer.java:322)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.IsolationAnalyzer.visit(IsolationAnalyzer.java:431)
        at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:76)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.IsolationAnalyzer.analyzeNode(IsolationAnalyzer.java:322)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.IsolationAnalyzer.visit(IsolationAnalyzer.java:368)
        at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:167)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.IsolationAnalyzer.analyze(IsolationAnalyzer.java:331)
        at io.ballerina.projects.internal.CompilerPhaseRunner.isolationAnalyze(CompilerPhaseRunner.java:204)
        at io.ballerina.projects.internal.CompilerPhaseRunner.performTypeCheckPhases(CompilerPhaseRunner.java:129)
        at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:435)
        at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
        at io.ballerina.projects.ModuleContext.compile(ModuleContext.java:383)
        at io.ballerina.projects.PackageCompilation.compileModulesInternal(PackageCompilation.java:208)
        at io.ballerina.projects.PackageCompilation.compileModules(PackageCompilation.java:192)
        at io.ballerina.projects.PackageCompilation.compile(PackageCompilation.java:99)
        at io.ballerina.projects.PackageCompilation.from(PackageCompilation.java:94)
        at io.ballerina.projects.PackageContext.getPackageCompilation(PackageContext.java:242)
        at io.ballerina.projects.Package.getCompilation(Package.java:150)
        at io.ballerina.projects.Package.runCodeGeneratorPlugins(Package.java:319)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:142)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.BuildCommand.execute(BuildCommand.java:288)
        at java.base/java.util.Optional.ifPresent(Optional.java:178)
        at io.ballerina.cli.launcher.Main.main(Main.java:58)
 
ERROR [issues:(1:1,1:1)] Compilation failed due to: Index 0 out of bounds for length 0
error: compilation contains errors
```

### Steps to Reproduce

```ballerina
public isolated function last(any v, any... vs) {
}

public function main() {
     _ = last(...[]);
}
```

### Affected Version(s)

2201.8.3

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by LakshanWeerasinghe**
Fixes #41691
Fixes #41766
Fixes #41818

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java [hunk 1]
```diff
@@ -1618,6 +1618,11 @@ public void visit(BLangListConstructorExpr listConstructorExpr) {
         }
     }
 
+    @Override
+    public void visit(BLangListConstructorExpr.BLangListConstructorSpreadOpExpr listConstructorSpreadOpExpr) {
+        analyzeNode(listConstructorSpreadOpExpr.expr, env);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:1623]: Can't we get rid of the special in the `BLangListConstructorExpr` visit with this?

**Review by MaryamZi — COMMENTED** [review]
Seems like the warning is now logged for 

```ballerina
class Foo {
    private int i = 1;

    function init() {
        
    }
}

service on new Listener() {
    resource function default test() {
        _ = start fn(new Foo());
    }
}

function fn(Foo foo) {
    
}

public isolated class Listener {

    public function 'start() returns error? {
    }

    public function gracefulStop() returns error? {
    }

    public function immediateStop() returns error? {
    }

    public function detach(service object {} s) returns error? {
    }

    public function attach(service object {} s, string[]|string|() name = ()) returns error? {
    }
}
```

**Comment by MaryamZi** [comment]
> Seems like the warning is now logged for
> 
> ```ballerina
> class Foo {
>     private int i = 1;
> 
>     function init() {
>         
>     }
> }
> 
> service on new Listener() {
>     resource function default test() {
>         _ = start fn(new Foo());
>     }
> }
> 
> function fn(Foo foo) {
>     
> }
> 
> public isolated class Listener {
> 
>     public function 'start() returns error? {
>     }
> 
>     public function gracefulStop() returns error? {
>     }
> 
>     public function immediateStop() returns error? {
>     }
> 
>     public function detach(service object {} s) returns error? {
>     }
> 
>     public function attach(service object {} s, string[]|string|() name = ()) returns error? {
>     }
> }
> ```

The warning is still logged for this, right?

**Comment by MaryamZi** [comment]
> > Seems like the warning is now logged for
> > ```ballerina
> > class Foo {
> >     private int i = 1;
> > 
> >     function init() {
> >         
> >     }
> > }
> > 
> > service on new Listener() {
> >     resource function default test() {
> >         _ = start fn(new Foo());
> >     }
> > }
> > 
> > function fn(Foo foo) {
> >     
> > }
> > 
> > public isolated class Listener {
> > 
> >     public function 'start() returns error? {
> >     }
> > 
> >     public function gracefulStop() returns error? {
> >     }
> > 
> >     public function immediateStop() returns error? {
> >     }
> > 
> >     public function detach(service object {} s) returns error? {
> >     }
> > 
> >     public function attach(service object {} s, string[]|string|() name = ()) returns error? {
> >     }
> > }
> > ```
> 
> The warning is still logged for this, right?



Still seeing the warning for this.

**Comment by LakshanWeerasinghe** [comment]
> > > Seems like the warning is now logged for
> > > ```ballerina
> > > class Foo {
> > >     private int i = 1;
> > > 
> > >     function init() {
> > >         
> > >     }
> > > }
> > > 
> > > service on new Listener() {
> > >     resource function default test() {
> > >         _ = start fn(new Foo());
> > >     }
> > > }
> > > 
> > > function fn(Foo foo) {
> > >     
> > > }
> > > 
> > > public isolated class Listener {
> > > 
> > >     public function 'start() returns error? {
> > >     }
> > > 
> > >     public function gracefulStop() returns error? {
> > >     }
> > > 
> > >     public function immediateStop() returns error? {
> > >     }
> > > 
> > >     public function detach(service object {} s) returns error? {
> > >     }
> > > 
> > >     public function attach(service object {} s, string[]|string|() name = ()) returns error? {
> > >     }
> > > }
> > > ```
> > 
> > 
> > The warning is still logged for this, right?
> 
> Still seeing the warning for this.

I will check this.

**Comment by LakshanWeerasinghe** [comment]
> > > Seems like the warning is now logged for
> > > ```
> > > class Foo {
> > >     private int i = 1;
> > > 
> > >     function init() {
> > >         
> > >     }
> > > }
> > > 
> > > service on new Listener() {
> > >     resource function default test() {
> > >         _ = start fn(new Foo());
> > >     }
> > > }
> > > 
> > > function fn(Foo foo) {
> > >     
> > > }
> > > 
> > > public isolated class Listener {
> > > 
> > >     public function 'start() returns error? {
> > >     }
> > > 
> > >     public function gracefulStop() returns error? {
> > >     }
> > > 
> > >     public function immediateStop() returns error? {
> > >     }
> > > 
> > >     public function detach(service object {} s) returns error? {
> > >     }
> > > 
> > >     public function attach(service object {} s, string[]|string|() name = ()) returns error? {
> > >     }
> > > }
> > > ```
> > 
> > 
> > The warning is still logged for this, right?
> 
> Still seeing the warning for this.

This behavior was visible when we do a `bal run`. But it won't appear when we do a `bal build`. The reason was the order of the entries coming from the `IsolationInferableHashMap` was different in the two scenarios. As per the offline discussion we had the order shouldn't matter when inferring the isolatedness of a function. When further investigating I found we are not inferring the isolation of dependent function call argument expression. After call the `isIsolatedExpression` with correct arguments to infer the expression, got the correct result.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java [hunk 2]
```diff
@@ -2308,112 +2338,214 @@ private void analyzeArgIsolatedness(BLangInvocation invocationExpr, List<BLangEx
...
-                    if (!Symbols.isFlagOn(params.get(i).flags, Flags.ISOLATED_PARAM)) {
-                        if (listConstrVarArg) {
-                            analyzeNode(listConstructorExpr.exprs.get(tupleIndex), env);
+    private void analyzeMappingConstructorRestArgs(BLangInvocation invocationExpr, BInvokableSymbol symbol,
+                                                boolean expectsIsolation, List<BVarSymbol> params, int paramsCount,
+                                                int reqArgCount, BLangRestArgsExpression varArg, BType varArgType,
+                                                Location varArgPos) {
+        BLangExpression varArgExpr = varArg.expr;
+        if (varArgType.tag != TypeTags.RECORD) {
+            return;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:2371]: There seem to be multiple codecov warnings?

Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:2362]: This check is to be on the safe side for all cases. Anyway, compiler won't allow mapping constructor values other than record type to pass as function arguments.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java [hunk 3]
```diff
@@ -293,10 +295,13 @@ public class IsolationAnalyzer extends BLangNodeVisitor {
 
     private boolean inferredIsolated = true;
     private boolean inLockStatement = false;
+    private boolean inIsolationStartAction = false;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:298]: Isolation -> Isolated?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java [hunk 4]
```diff
@@ -2308,112 +2333,160 @@ private void analyzeArgIsolatedness(BLangInvocation invocationExpr, List<BLangEx
...
+                            break;
+                        }
+                    } else if (recordField.getKind() == NodeKind.SIMPLE_VARIABLE_REF) {
+                        BLangRecordLiteral.BLangRecordVarNameField varRefRecordField = 
+                                ((BLangRecordLiteral.BLangRecordVarNameField) recordField);
+                        if (varRefRecordField.symbol.name.value.equals(param.name.value)) {
+                            arg = varRefRecordField;
+                            break;
+                        }
+                    } else {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:2396]: There still seem to be codecov warnings.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java (hunk 2 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:2362]: We can go ahead and cast then.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java [hunk 5]
```diff
@@ -1134,14 +1139,25 @@ public void visit(BLangRecordLiteral recordLiteral) {
...
+                analyzeNode((BLangRecordLiteral.BLangRecordSpreadOperatorField) field, env);
             }
         }
     }
 
+    @Override
+    public void visit(BLangRecordLiteral.BLangRecordSpreadOperatorField spreadOperatorField) {
+        analyzeNode(spreadOperatorField.expr, env);
+        NodeKind kind = spreadOperatorField.expr.getKind();
+        if (!(kind == NodeKind.RECORD_LITERAL_EXPR || kind == NodeKind.TYPE_CONVERSION_EXPR)) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:1156]: What is this special case we're trying to handle here? Why do we need to special case the type cast expression?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java [hunk 6]
```diff
@@ -1270,9 +1286,12 @@ public void visit(BLangSimpleVarRef varRefExpr) {
                 }
             }
         }
-
         if (!recordFieldDefaultValue && !objectFieldDefaultValueRequiringIsolation && enclInvokable != null &&
                 isReferenceToVarDefinedInSameInvokable(symbol.owner, enclInvokable.symbol)) {
+            if (this.inIsolationStartAction
+                    && !isSubtypeOfReadOnlyOrIsolatedObjectOrInferableObject(symbol.owner, symbol.getType())) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:1292]: Do we have tests where a normal function call is an arg to an isolated start action?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java (hunk 5 above)
Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:1156]: In here we only need to add whats inside the mapping constructor (key values and simple var refs).

Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:1156]: Inside the spread operator field, the expression can be type cast expression.

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:1156]: Spread (with a mapping constructor) can be other expressions too, not just these?

Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:1156]: Only these will be allowed to pass as a function argument. Others will be not allowed by the type checking phase.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java [hunk 7]
```diff
@@ -2308,112 +2333,152 @@ private void analyzeArgIsolatedness(BLangInvocation invocationExpr, List<BLangEx
...
 
-                int tupleIndex = 0;
-                for (int i = reqArgCount; i < paramsCount; i++) {
-                    if (!Symbols.isFlagOn(params.get(i).flags, Flags.ISOLATED_PARAM)) {
-                        if (listConstrVarArg) {
-                            analyzeNode(listConstructorExpr.exprs.get(tupleIndex), env);
+    private void analyzeMappingConstructorRestArgs(BLangInvocation invocationExpr, BInvokableSymbol symbol,
+                                                boolean expectsIsolation, List<BVarSymbol> params, int paramsCount,
+                                                int reqArgCount, BLangRestArgsExpression varArg, BType varArgType,
+                                                Location varArgPos) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:2359]: There are some parameters here that are unused.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java (hunk 5 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:1156]: Is that statement correct? 

```ballerina
type Record record {|
    string[] arr1;
    string[] arr2;
    string[] arr3;
|};

function qux() {
    f(...<Record> {arr1: [], arr2: [], ...({arr3: []})}); // grouped expression in spread
}

isolated function f(string[] arr1, string[] arr2, string[] arr3) {
}
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java (hunk 4 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:2393]: Still seeing codecov warnings.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java (hunk 6 above)
Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java]: added in https://github.com/ballerina-platform/ballerina-lang/pull/41772/commits/3f069ce069ced6817f05c87fb0c77f5f82867da4

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java (hunk 5 above)
Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java]: I didn't figure out this is possible previously. Handle it in [commit](https://github.com/ballerina-platform/ballerina-lang/pull/41772/commits/d7e77290ba3bfc7bf08e9e6703fee8cc132aeea9).

**Pull request closed without merging**